### PR TITLE
Add inference program for Transformer.

### PR DESCRIPTION
**Pull request by guoshengCS**
Add inference program for Transformer.

### CODE REVIEW

File: fluid/neural_machine_translation/transformer/config.py [hunk 1]
```diff
@@ -15,6 +15,23 @@ class TrainTaskConfig(object):
     # the params for learning rate scheduling
     warmup_steps = 4000
 
+    # the directory for saving inference models
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: for saving inference models --> for saving trained models

File: fluid/neural_machine_translation/transformer/config.py [hunk 2]
```diff
@@ -15,6 +15,23 @@ class TrainTaskConfig(object):
...
+    model_dir = "transformer_model"
+
+
+class InferTaskConfig(object):
+    use_gpu = False
+    # number of sequences contained in a mini-batch
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: - number of sequences contained in a mini-batch --> the number of examples in one run for sequence generation.
- Please add a comment here to warn users currently the batch size can only be set to 1.

Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: number of sequences contained in a mini-batch --> the number of examples in one run for sequence generation.

File: fluid/neural_machine_translation/transformer/config.py [hunk 3]
```diff
@@ -15,6 +15,23 @@ class TrainTaskConfig(object):
...
+    batch_size = 1
+
+    # the params for beam search
+    beam_size = 5
+    max_length = 30
+    n_best = 1
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: please comment `n_best`. It is confusing to me about what is the difference between `beam_size` and `n_best`.

File: fluid/neural_machine_translation/transformer/config.py [hunk 4]
```diff
@@ -15,6 +15,23 @@ class TrainTaskConfig(object):
     # the params for learning rate scheduling
     warmup_steps = 4000
 
+    # the directory for saving inference models
+    model_dir = "transformer_model"
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: change the name to "trained_models"

File: fluid/neural_machine_translation/transformer/config.py [hunk 5]
```diff
@@ -15,6 +15,23 @@ class TrainTaskConfig(object):
...
+
+
+class InferTaskConfig(object):
+    use_gpu = False
+    # number of sequences contained in a mini-batch
+    batch_size = 1
+
+    # the params for beam search
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: params --> parameters, please avoid using an unnecessary abbreviation.

File: fluid/neural_machine_translation/transformer/config.py [hunk 6]
```diff
@@ -15,6 +15,23 @@ class TrainTaskConfig(object):
...
+    beam_size = 5
+    max_length = 30
+    n_best = 1
+
+    # the directory for loading inference model
+    model_path = "transformer_model/pass_1.infer.model"
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/config.py]: please rename the directory to "trained_models"

File: fluid/neural_machine_translation/transformer/infer.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        trg_words = np.array(
+            [[bos_idx]] * batch_size * beam_size, dtype="int64")
+        trg_pos = np.array([[1]] * batch_size * beam_size, dtype="int64")
+        src_max_length, src_slf_attn_bias, trg_max_len = enc_in_data[
+            -1], enc_in_data[-2], 1
+        trg_src_attn_bias = np.tile(
+            src_slf_attn_bias[:, :, ::src_max_length, :],
+            [beam_size, 1, trg_max_len, 1])
+        enc_output = np.tile(enc_output, [beam_size, 1, 1])
+        # No need for trg_slf_attn_bias because of no paddings.
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/infer.py]: this comment is hard to understand.

File: fluid/neural_machine_translation/transformer/model.py [hunk 8]
```diff
@@ -373,6 +376,53 @@ def decoder(dec_input,
...
 
 
+def make_inputs(input_data_names,
+                n_head,
+                d_model,
+                batch_size,
+                max_length,
+                slf_attn_bias_flag,
+                src_attn_bias_flag,
+                pos_flag=1):
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/model.py]: change "pos_flag" into "is_pos=True"

File: fluid/neural_machine_translation/transformer/model.py [hunk 9]
```diff
@@ -387,61 +437,72 @@ def transformer(
...
-    # The actual shape of trg_src_attn_bias in runtime is:
-    # [batch_size, n_head, max_trg_length_in_batch, max_src_length_in_batch].
-    # This is used to remove attention weights on paddings.
-    trg_src_attn_bias = layers.data(
-        name=input_data_names[6],
-        shape=[batch_size, n_head, max_length, max_length],
-        dtype="float32",
-        append_batch_size=False)
+    enc_input_layers = make_inputs(encoder_input_data_names, n_head, d_model,
+                                   batch_size, max_length, 1, 0)
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/model.py]: 0 --> False

File: fluid/neural_machine_translation/transformer/model.py [hunk 10]
```diff
@@ -387,61 +437,72 @@ def transformer(
...
+        d_model,
+        d_inner_hid,
+        dropout_rate,
+        src_pad_idx,
+        pos_pad_idx,
+        enc_input_layers, )
+
+    dec_input_layers = make_inputs(decoder_input_data_names, n_head, d_model,
+                                   batch_size, max_length, 1, 1)
+
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/model.py]: the last two 1 --> True

File: fluid/neural_machine_translation/transformer/model.py [hunk 11]
```diff
@@ -387,61 +437,72 @@ def transformer(
...
+        dropout_rate,
+        trg_pad_idx,
+        pos_pad_idx,
+        dec_input_layers,
+        enc_output, )
+
+    # Padding index do not contribute to the total loss. The weights is used to
+    # cancel padding index in calculating the loss.
+    gold, weights = make_inputs(label_data_names, n_head, d_model, batch_size,
+                                max_length, 0, 0, 0)
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/model.py]: make the last three parameters of `make_inputs` boolean parameters.

File: fluid/neural_machine_translation/transformer/model.py [hunk 12]
```diff
@@ -373,6 +376,53 @@ def decoder(dec_input,
...
+        dtype="int64" if pos_flag else "float32",
+        append_batch_size=False)
+    input_layers += [pos]
+    if slf_attn_bias_flag:
+        # This is used for attention bias or encoder output.
+        slf_attn_bias = layers.data(
+            name=input_data_names[2]
+            if slf_attn_bias_flag == 1 else input_data_names[-1],
+            shape=[batch_size, n_head, max_length, max_length]
+            if slf_attn_bias_flag == 1 else [batch_size, max_length, d_model],
```
Comment (lcy-seso) [inline fluid/neural_machine_translation/transformer/model.py]: make `src_attn_bias_flag` a boolean parameter.

**Review by lcy-seso — COMMENTED** [review]
Thank you for this work.

File: fluid/neural_machine_translation/transformer/config.py (hunk 1 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

File: fluid/neural_machine_translation/transformer/config.py (hunk 4 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

File: fluid/neural_machine_translation/transformer/config.py (hunk 5 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

File: fluid/neural_machine_translation/transformer/config.py (hunk 2 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

File: fluid/neural_machine_translation/transformer/config.py (hunk 3 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

File: fluid/neural_machine_translation/transformer/config.py (hunk 6 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/config.py]: Done.

File: fluid/neural_machine_translation/transformer/infer.py (hunk 7 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/infer.py]: Fixed. Actually `trg_slf_attn_bias` is needed in decoder for inference.

File: fluid/neural_machine_translation/transformer/model.py (hunk 12 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/model.py]: Done.

File: fluid/neural_machine_translation/transformer/model.py (hunk 8 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/model.py]: Done.

File: fluid/neural_machine_translation/transformer/model.py (hunk 9 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/model.py]: Done.

File: fluid/neural_machine_translation/transformer/model.py (hunk 10 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/model.py]: Done.

File: fluid/neural_machine_translation/transformer/model.py (hunk 11 above)
Comment (guoshengCS) [reply fluid/neural_machine_translation/transformer/model.py]: Done.

**Review by lcy-seso — APPROVED** [review]
LGTM.

**Pull request merged**